[doc][yba] xCluster Replication update (#23417)
yugabyte/yugabyte-db: docs/content/preview/yugabyte-platform/manage-deployments/xcluster-replication/bidirectional-replication.md

type: docs
---

With bidirectional replication, both universes can perform reads and writes, and writes are replicated in both directions. This is also referred to as [Active-active multi-master](../../../architecture/docdb-replication/async-replication/#active-active-multi-master).

You set up bidirectional replication using YugabyteDB Anywhere by creating two separate non-transactional xCluster replication configurations. Under this scenario, you create a replication configuration from universe A to universe B, and then you create a new replication configuration from universe B to universe A.

YugabyteDB Anywhere detects an xCluster setup is for bidirectional if the tables that are being added to the replication are already part of a replication in the reverse direction (from A to B). In this scenario, YugabyteDB Anywhere skips the full copy for tables on universe B that are already in replication.

{{<note title="Bidirectional xCluster deployment">}}
Due to its operational complexity, bidirectional xCluster is not recommended for most use cases. If you are considering a bidirectional xCluster deployment, contact {{% support-general %}}.
{{</note>}}

## Limitations

> - Bidirectional replication can only be done using non-transactional replication. Transactional replication is not supported because transactional replication puts the target universe in read-only mode. For more information, see [Asynchronous replication modes](../../../../architecture/docdb-replication/async-replication/#asynchronous-replication-modes).

Reviewer: Read only mode is not the reason for it.
It's not supposed since xcluster is async with last writer wins semantics. That is not transactional or ACID compliant.

Author: @shahrooz1997

Reviewer 2: I agree that mentioning read-only convey more useful information to the user rather than explaining why it was decided to go with this approach.

- You can't use the YugabyteDB Anywhere UI to create two separate replication configurations for YSQL, each containing a subset of the database tables.

> - Active-active bidirectional replication is not supported because the backup or restore would wipe out the existing data. This means that copying data can be done only if an xCluster configuration with reverse direction for a table does not exist. It is recommended to set up replication from your active universe to the passive target, and then set up replication for the target to the source universe. To restart a replication with a full copy, the reverse replication must be deleted.

Reviewer: Bi-directional is the very definition of active-active.
And YBA supports bi-directional. What's the intent of this?

Author: @shahrooz1997

Reviewer 2: @hari90 >> And YBA supports bi-directional. What's the intent of this?
We can improve the wording to reflect better the intention. If for any reason the replication breaks, you will have data loss to set up the replication again through YBA, unless to manually merge the data on one side and set up replication then.

@iSignal Makes sense. We could remove this paragraph and explain it in the restart section of bidirectional.

You set up bidirectional replication using YugabyteDB Anywhere by creating two separate non-transactional xCluster replication configurations. Under this scenario, you create a replication configuration from universe A to universe B, and then you create a new replication configuration from universe B to universe A.

![Active-Active Multi-Master](/images/architecture/replication/active-active-deployment-new.png)

YugabyteDB Anywhere detects an xCluster setup is for bidirectional if the tables that are being added to the replication are already part of a replication in the reverse direction (from A to B). In this scenario, YugabyteDB Anywhere skips the full copy for tables on universe B that are already in replication.

{{<note title="Bidirectional xCluster deployment">}}
Due to its operational complexity, bidirectional xCluster is not recommended for most use cases. If you are considering a bidirectional xCluster deployment, contact {{% support-general %}}.
{{</note>}}

## Limitations

> - Bidirectional replication can only be done using non-transactional replication. Transactional replication is not supported because transactional replication puts the target universe in read-only mode. For more information, see [Asynchronous replication modes](../../../../architecture/docdb-replication/async-replication/#asynchronous-replication-modes).

Reviewer 3: ```suggestion
- Bidirectional replication can only be done using non-transactional replication. For more information, see [Asynchronous replication modes](../../../../architecture/docdb-replication/async-replication/#asynchronous-replication-modes).
```

## Prerequisites

- Create two universes as described in [Prerequisites](../xcluster-replication-setup/#prerequisites).

## Set up bidirectional replication

To set up bidirectional replication from universe A to universe B, and vice versa on the same set of tables or databases, do the following:

1. Make sure that there are no writes happening on universe B.
1. Set up xCluster replication from A to B following the steps in [Deploy xCluster](../../../../deploy/multi-dc/async-replication/async-deployment/). A full copy of the selected tables or databases may be performed as part of this operation.
1. Set up xCluster replication from universe B to A following the steps in [Deploy xCluster](../../../../deploy/multi-dc/async-replication/async-deployment/).

    No full copy is performed, so it is important to have no writes on B during this setup, as these writes might not be replicated to universe A.

For information on how to monitor and manage the replication configurations, refer to [Monitor replication](../xcluster-replication-setup/#monitor-replication).

## Restart bidirectional replication

> Restarting bidirectional replication can cause unreplicated data on one of the universes to be lost, because a full copy is performed as part of the restart. Copying data can be done only if an xCluster configuration with reverse direction for a table does not exist, which means the reverse replication must first be deleted.

Reviewer 3: ```suggestion
In case of an extended network partition or replication is broken due to other reasons, it might be necessary to restart replication. However, restarting replication on individual directions of a bidirectional configuration does not achieve this goal.
```

> As a result, before restarting bidirectional replication, you need to identify the universe whose data is more up to date.

Reviewer 3: ```suggestion
Before restarting bidirectional replication, you need to identify the universe whose data is more up to date.
```

To restart a bidirectional replication setup:

1. Identify the universe whose data is more up to date.
1. Delete the replication configuration where this universe is the target universe.
1. Restart the other replication configuration, where this universe is the source universe.
1. Set up the replication configuration deleted in step (2) again, with this universe as the target universe.

Note that restarting one side of an existing bidirectional replication configuration without performing step 2 simply re-establishes the replication configuration from the current data without performing a full copy. This can cause inconsistencies in the data between the source and target universes.

## DDL operations in bidirectional replication

In general, for DDL operations, you can follow the steps in [Manage tables and indexes](../xcluster-replication-ddl/) while taking care to update the xCluster replication configuration on both configurations involved.

However, in some special cases, you need to follow different procedures.

> 

Reviewer 3: ### Add a database to bidirectional xcluster replication

It is recommended that the databases should be added to bidirectional configurations right after they are created before any writes are performed on the tables within the dbs. 

To add a newly created database to bidirectional replication, follow the instructions at [Add a database to xcluster replication] in both directions.

Non-empty databases can still be added to replication in both directions but note that YBA will perform a [full copy](link to full copy section) to the target universe when it is added to the first direction, causing the current database on the target to be re-created.

You set up bidirectional replication using YugabyteDB Anywhere by creating two separate non-transactional xCluster replication configurations. Under this scenario, you create a replication configuration from universe A to universe B, and then you create a new replication configuration from universe B to universe A.

![Active-Active Multi-Master](/images/architecture/replication/active-active-deployment-new.png)

YugabyteDB Anywhere detects an xCluster setup is for bidirectional if the tables that are being added to the replication are already part of a replication in the reverse direction (from A to B). In this scenario, YugabyteDB Anywhere skips the full copy for tables on universe B that are already in replication.

{{<note title="Bidirectional xCluster deployment">}}
Due to its operational complexity, bidirectional xCluster is not recommended for most use cases. If you are considering a bidirectional xCluster deployment, contact {{% support-general %}}.
{{</note>}}

## Limitations

In addition to the regular [limitations](../#limitations), Bidirectional replication can only be done using [non-transactional replication](../#xcluster-configurations).

## Prerequisites

> Create two universes as described in [Prerequisites](../xcluster-replication-setup/#prerequisites).

Reviewer 4: ```suggestion
Create two universes as described in set up [Prerequisites](../xcluster-replication-setup/#prerequisites).
```
